distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala

```diff
@@ -1114,4 +1114,76 @@ object Lifecycle {
 
   @deprecated("renamed to fromAutoCloseable", "0.11")
   def fromAutoCloseableF[F[_], A <: AutoCloseable](acquire: => F[A])(implicit F: DIEffect[F]): Lifecycle[F, A] = fromAutoCloseable(acquire)
+
+
+  /**
+    * This type encoding was copy-pasted from cats.effect.Resource
+    * It's needed to declare a new type with cats.Applicative
+    * capabilities of doing parallel processing in `ap` and `map2`, needed
+    * for implementing `cats.Parallel`.
+    */
+  type Par[+F[_], +A] = Par.Type[F, A]
+  object Par {
+    type Base
+    trait Tag extends Any
+    type Type[+F[_], +A] <: Base with Tag
+
+    def apply[F[_], A](fa: Lifecycle[F, A]): Type[F, A] =
+      fa.asInstanceOf[Type[F, A]]
+
+    def unwrap[F[_], A](fa: Type[F, A]): Lifecycle[F, A] =
+      fa.asInstanceOf[Lifecycle[F, A]]
+  }
+}
+
+abstract private[definition] class LifecycleCatsInstances0 {
+  import izumi.fundamentals.orphans._
+
+  implicit def catsMonoidForLifecycle[F[_], Monoid[_]: `cats.Monoid`, S[_[_]]: `cats.effect.Sync`, Ap[_[_]]: `cats.Applicative`, SG[_]: `cats.Semigroup`, A](
+    implicit
+    Monoid: Monoid[F[_]],
+    S: S[F],
+    Ap: Ap[F],
+    SG: SG[A],
```

Comment:
Feels like you don't need both Semigroup for A and Monoid[F[A]]. Maybe just use `Monoid[A]` and implement `empty` as `Lifecycle.pure(M.empty)` ?